### PR TITLE
Satellite Managed Encryption Projects

### PR DESCRIPTION
**Pull request by mobyvb**
Design document for projects created with satellite managed encryption
https://github.com/storj/storj/issues/6620

### CODE REVIEW

File: 20240129-satellite-managed-encryption.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+tags: []
```
Comment (ifraixedes) [inline 20240129-satellite-managed-encryption.md:1]: Please add some tags

Comment (ifraixedes) [inline 20240129-satellite-managed-encryption.md:1]: It would be nice to follow the template https://github.com/storj/design-docs/blob/main/00000000-template.md

There is a document that led to creating it https://storjlabs.atlassian.net/wiki/spaces/delivery/pages/2548137985/Design+Doc+Process+Proposal+2023

Unfortunately, that Confluence page isn't public, we should make it public, so people outside of Storj can find out why we have the template, etc.

File: 20240129-satellite-managed-encryption.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+## Design
+
+1. add a new `projects` DB column, like `projects.satellite_managed_encryption`, with all existing and new projects having this turned off
+2. Key Management Store on satellite:
+    * a secure way to associated project IDs with random encryption passphrases
+    * should have heavy access restrictions and/or be encrypted
+    * should be just as reliable as metainfo DB, since data loss or corruption from the KMS means data lost on the network
+3. Satellite GUI changes:
+    * support ability to create a project with "satellite managed encryption" turned on
```
Comment (elek) [inline 20240129-satellite-managed-encryption.md:35]: Do we expect a separated design doc about this, or extending this one with more information?

File: 20240129-satellite-managed-encryption.md (hunk 1 above)
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: done

Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: done

File: 20240129-satellite-managed-encryption.md (hunk 2 above)
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: I updated the document with a different template, and filled out the UI section a little more, so you may want to take another look.
However, I do not think it is necessary to elaborate further on the UI. Once we define the core requirements, Tome will provide the Console team with info about how the UI should look for project and access creation.
I don't anticipate the changes in the UI to be very dramatic. The main difference will be:
* ability to select between "satellite-managed" or "user-managed" encryption during project creation
* never prompted for encryption passphrase for satellite-managed encryption project
* access credentials updated with correct encryption behind-the-scenes in the UI, for satellite-managed encryption project

Comment (elek) [reply 20240129-satellite-managed-encryption.md]: Still I don't fully understand what is Satellite KMS. From my point of view, that's the key of this change. This document is either a UI only design doc (in this case we need a separated one for Satellite KMS) or a full (my preference), but in this case the Satellite KMS should also be covered with more details.

I like the vision from JT: "storj-managed keys should be at least as secure as AWS [/GCE] managed keys". 

Also it was mentioned somewhere, that database leak shouldn't compromise all the managed secrets.

I think it would be important to share the architecture of the key management, and defined what can be expected.

File: 20240129-satellite-managed-encryption.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+We need a secure way to store and look up project encryption passphrases on the Satellite. Someone with normal Satellite DB access should not be able to view this data, and ideally it is encrypted. At the same time, it must be very reliable, as losing data about project encryption passphrases means that data may have been lost.
+
+The specific technology/product used for the KMS can be decided in a later stage. The primary features and concerns we should keep in mind:
+* as simple as possible from an implementation perspective, e.g. key-value lookup for project ID -> encryption passphrase
+* secure - should be encrypted, and highly restricted who/what has access to it
+* reliable - should have the same standards for uptime/data loss as metainfo DB
+
+#### Satellite Project Updates
+
+A new flag should be added to the `projects` table, like `projects.satellite_managed_encryption`. This should be false for all existing projects. While satellite-managed-encryption (this design) is in-progress, new projects should be created with this flag set to false.
```
Comment (egonelbre) [inline 20240129-satellite-managed-encryption.md:56]: Do we want to be able to later customize this based on the bucket?

File: 20240129-satellite-managed-encryption.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the new column in the `projects` table is set to `true` for this project
+* a new key-value pair is created for the new project in the Key Management Store. The passphrase should be cryptographically random
+* endpoints on the satellite API are added or updated so that an authenticated user can request the encryption passphrase for a project during access creation
+
+#### WebAssembly Updates
+
+Wasm functionality must be updated to allow the Satellite UI to properly generate access credentials with path encryption disabled. [Code for reference](https://github.com/storj/storj/blob/a62929fd5757a40eda5d0b044ad2cefc14708410/satellite/console/consolewasm/access.go#L22-L29).
```
Comment (egonelbre) [inline 20240129-satellite-managed-encryption.md:65]: Does https://github.com/storj/access-python need updating  as well?

File: 20240129-satellite-managed-encryption.md (hunk 3 above)
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:56]: as far as I know, no. From a product and design perspective, I think the intention is to generally pressure users (except highly technical/advanced users) into the idea that 1 project = 1 encryption passphrase

File: 20240129-satellite-managed-encryption.md (hunk 2 above)
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: Agreed. I'm discussing with others about what the details should be, but I plan to update this design rather than creating a new one.

File: 20240129-satellite-managed-encryption.md (hunk 4 above)
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:65]: Possibly. I guess I will also need to look at our other bindings too

File: 20240129-satellite-managed-encryption.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+The user-managed-encryption project will be the same as before, and user will always be prompted to enter encryption passphrase when necessary, e.g. when using the file browser or generating an access grant.
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
+
+The user should not be able to directly retrieve a "raw API key" from the Satellite UI. This prevents the user from unintentionally creating "invalid" credentials using a raw API key with Uplink. The only credentials that can be generated from the access page for a satellite-managed-encryption project should be credentials that already have embedded encryption rules.
+
+The file browser in the UI should generate access credentials using the passphrase from the satellite KMS, and no path encryption.
+
+#### Uplink Updates
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:82]: Yes it could, but I don't think these changes are completely necessary given the other changes described. If the UI only provides the user with a restricted access credential, it will be difficult for the user to accidentally setup uplink with the wrong encryption, even if we make no changes

[the remark this answers is not in the document]
Comment (ifraixedes) [reply 20240129-satellite-managed-encryption.md:82]: I meant, that we should mention that those would require updates as we are mentioning here that uplink-cli does.

File: 20240129-satellite-managed-encryption.md (hunk 2 above)
Comment (wilfred-asomanii) [reply 20240129-satellite-managed-encryption.md]: The doc has been updated with details about the KMS.

File: 20240129-satellite-managed-encryption.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+* provide users with an experience that does not require managing and entering an encryption passphrase in the Satellite UI
+* provide users with an easy way to view and list files in lexographical order
+
+### Approach / Design
+
+#### Key Management Store
+
+We need a secure way to store and look up project encryption passphrases on the Satellite. Someone with normal Satellite DB access should not be able to view this data, and ideally it is encrypted. At the same time, it must be very reliable, as losing data about project encryption passphrases means that data may have been lost.
+
+The specific technology/product used for the KMS can be decided in a later stage. The primary features and concerns we should keep in mind:
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: @wilfred-asomanii has updated the document with info about the KMS

File: 20240129-satellite-managed-encryption.md [hunk 7]
```diff
@@ -44,13 +44,34 @@ The main goals are to:
...
+For deployment, we have the option to host Vault/Infisical ourselves and manage it securely, or use a managed instances from Harshicorp/Infisical themselves. However, we choose to deploy it, we will need to ensure that it is highly available and secure so as not to leak or lose passphrases.
+We must make sure to restrict access to deployment secrets.
+
+To interact with the KMS, the satellite will have to be configured with;
+* The address of the KMS
+* Secret path (Where the secret is stored - [Vault](https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v1#path), It means the same for Infisical too)
+* The authentication token used to initialise the vault client
+* Or the Infisical access token and Infisical API key set in the `Authorization` and `X-API-KEY` header respectively in an http request.
+
+It is important that these config values be mre protected than what is currently in storj/infra.
```
Comment (ifraixedes) [inline 20240129-satellite-managed-encryption.md:73]: mre?

File: 20240129-satellite-managed-encryption.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ifraixedes) [reply 20240129-satellite-managed-encryption.md]: Vault sounds good and we have mentioned a few times that it would be nice to have to store all our secrets.

However,  in my opinion, deploying and integrating it in the satellite should have a separate design doc because of what's commented in this design doc about it.

[the remark this answers is not in the document]
Comment (ifraixedes) [reply 20240129-satellite-managed-encryption.md]: Vault isn't the selected option anymore.

File: 20240129-satellite-managed-encryption.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+* The KMS service - This is part of the Satellite that will be responsible for fetching the master key at startup, encrypting and decrypting passphrases, and storing and retrieving same from the passphrase store. The master key will be kept in memory throughout the lifetime of the Satellite.
+The satellite will be configured with credentials needed to access the Secret Manager (the master key store).
+* The Passphrase store - This where the encrypted passphrases will be stored. It will be a new column, `passphrase_enc` on the `projects` table.
+
+#### Satellite Project Updates
+
+A new flag should be added to the `projects` table, like `projects.satellite_managed_encryption`. This should be false for all existing projects. While satellite-managed-encryption (this design) is in-progress, new projects should be created with this flag set to false.
+Another column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and null for others.
```
Comment (kaloyan-raev) [inline 20240129-satellite-managed-encryption.md:62]: Why do we need to store the encrypted passphrase in the satellite DB? It's already stored in Google Secret Manager and can be accessed from there whenever necessary, doesn't it?

Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:62]: If only the "master key" is in Google Secret Manager, we just need to retrieve it once, on startup. Then we can decrypt anything in the `projects` table without needing to make any additional network requests - we just use the master key in memory.
Also, by having encrypted passphrase in the projects table, those encrypted passphrases will automatically be included in our Satellite DB backups. We can still back up Google Secret Manager, but with this implementation, we only need to worry about a single key rather than one for every project.

Comment (kaloyan-raev) [reply 20240129-satellite-managed-encryption.md:62]: As far as I remember, we discussed using the secret project ID as the satellite-managed passphrase. But then we decided to generate random passphrases and store them in a KMS because those with access to the satellite database shouldn't access the passphrases. Now it turns out we want only to store the master encryption key in the KMS and have the encrypted passphrases in the satellite DB...

If we only have the master encryption key in the KMS, how about storing the secret project ID encrypted in the satellite DB instead of in plaintext? And use it as the satellite-managed passphrase? This way, we won't need an additional `passphrase_enc` column.

Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:62]: I do not remember talking about using secret project ID in this way (are you just talking about `projects.id`, the primary key? The one that is not `projects.public_id`?)
Anyway, I don't think it matters where exactly the passphrase is stored, as long as:
* the raw passphrase is not accessible with only cockroach DB access
* the raw passphrase can easily be derived and/or decrypted by the satellite, given a secret stored outside of cockroach

It seems like there are several ways to do this. However, if we wanted to involve `project.id` instead of adding a new column like `passphrase_enc`, maybe the way to do it is to combine `project.id` + `KMS master key` (e.g. crypto hash function) to get the encryption passphrase. And it would not need to be stored anywhere, since it could be derived from data that is already stored.

Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:62]: I talked with Wilfred about this, and I think there is one potential problem with deriving the passphrase without having an additional column in the DB: rotating the master key.
If, for some reason, we wanted to change the master key in the KMS (e.g. because it was exposed), it is very straightforward if we are storing `passphrase_enc` in the DB:
* unencrypt `passphrase_enc` with old master key
* encrypt `passphrase` to `passphrase_enc` with new master key
* done
However, if the encryption passphrase is derived based on, for example, master key + project ID, the above process of migration is impossible. If the master key is leaked, the encryption passphrase is permanently leaked, because it is fundamentally derived from the master key.

In short:
* if we store `passphrase_enc` in the DB, we can change the master key without changing the underlying encryption passphrase of the project
* if we derive passphrase rather than storing it, we cannot change the master key unless we re-upload all objects in all the projects (which we will not do).

So for security reasons, I think it is preferable to store `passphrase_enc` in the DB

Comment (kaloyan-raev) [reply 20240129-satellite-managed-encryption.md:62]: Yes, I referred to `projects.id` as the secret project ID. But if it is the primary key, migrating the DB to encrypt it would be tricky.

I see your argument about rotating the master key and it makes sense.

File: 20240129-satellite-managed-encryption.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the `projects.satellite_managed_encryption` column is set to `true` for this project
```
Comment (kaloyan-raev) [inline 20240129-satellite-managed-encryption.md:65]: Do we need this column if we can check that `passphrase_enc` is `null`?

Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:65]: I don't think we need a separate boolean flag if we can use nullable `passphrase_enc` as a flag

Comment (wilfred-asomanii) [reply 20240129-satellite-managed-encryption.md]: I've removed that

File: 20240129-satellite-managed-encryption.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+This secrets manager was chosen because it is just as secure as the alternatives, if not more and yet cheaper and easier to manage. It also is more convenient to use as we already use GCP.
+* The KMS service - This is part of the Satellite that will be responsible for fetching the master key at startup, encrypting and decrypting passphrases, and storing and retrieving same from the passphrase store. The master key will be kept in memory throughout the lifetime of the Satellite.
+The satellite will be configured with credentials needed to access the Secret Manager (the master key store).
+* The Passphrase store - This where the encrypted passphrases will be stored. It will be a new column, `passphrase_enc` on the `projects` table. As added benefits to storing this on `projects`, we will be able to back up encrypted passphrases and also rotate master keys.
```
Comment (mobyvb) [inline 20240129-satellite-managed-encryption.md:57]: nit - I don't think we need to think of "the passphrase store" as its own thing. Above, we have "The KMS service ... will be responsible for ... encrypting and decryptiong passphrases, and storing and retrieving from the passphrase store"

I think the KMS maybe doesn't even need to be thought of as a "store", but rather just a simple "encryption/decryption service", whose primary responsibilities are:
* keep the master key in memory
* expose, e.g. `managedPassphraseService.GenerateNewPassphrase()` and `managedPassphraseService.Decrypt(encryptedPassphrase)` functionality

In other words, the "managed passphrase service" doesn't even need to have DB access - we are storing `passphrase_enc` in the DB, and this encrypted passphrase will already be accessible from the satellite/console package - all we need to be able to do is decrypt it, which is a very simple operation.

One reason I think this slight change would help:
Imagine we are creating a new project - we call db.CreateProject and pass in all the initial field values. This is happening from the Console service. That means that ideally, the Console service needs to have the encrypted project passphrase so that it can insert `passphrase_enc` at the same time the project is created. Otherwise, project creation and setting the satellite managed encryption passphrase need to happen as two separate DB queries.

File: 20240129-satellite-managed-encryption.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+However, there are [security considerations](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp?tab=readme-ov-file#security-considerations) we have to make.
+
+#### Satellite Project Updates
+
+A new column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and `null` for others.
+Projects that are created before this feature is deployed and projects created after but opt out should have `null` in this column.
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the KMS service generates a cryptographically random passphrase and encrypts it using the master key
+* the passphrase store is updated with the encrypted passphrase for this project
```
Comment (mobyvb) [inline 20240129-satellite-managed-encryption.md:79]: related to my other comment - here, I think it would be best to get the encrypted passphrase from the Key Management Service, and then pass the encrypted passphrase into the call to `db.CreateProject` _outside_ of the Key Management Service (minimizes the need to restructure code, I think)

File: 20240129-satellite-managed-encryption.md [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
+
+The user should not be able to directly retrieve a "raw API key" from the Satellite UI. This prevents the user from unintentionally creating "invalid" credentials using a raw API key with Uplink. The only credentials that can be generated from the access page for a satellite-managed-encryption project should be credentials that already have embedded encryption rules.
+
+The file browser in the UI should generate access credentials using the passphrase from the satellite KMS, and no path encryption.
+
+#### External Updates
+
+###### WebAssembly
```
Comment (mobyvb) [inline 20240129-satellite-managed-encryption.md:100]: nit - I would not consider the wasm code updates to be an "external update", since it is effectively served and executed from the Satellite. I would probably group the wasm updates with the "satellite UI" updates above, since they probably need to be done together

File: 20240129-satellite-managed-encryption.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+* Wasm functionality must be updated to allow the Satellite UI to properly generate access credentials with path encryption disabled. [Code for reference](https://github.com/storj/storj/blob/a62929fd5757a40eda5d0b044ad2cefc14708410/satellite/console/consolewasm/access.go#L22-L29).
+
+###### Uplink and Bindings
```
Comment (mobyvb) [inline 20240129-satellite-managed-encryption.md:104]: We should make it clear that this section is not 100% necessary. We can choose not to update Uplink and bindings, and as long as the UI is implemented correctly, nothing should break.

I think either:
* this heading should be prominently marked as "optional", or
* this section should be moved out of the main design section, and into a "possible future improvements" section (or even to "out of scope" at the bottom)

File: 20240129-satellite-managed-encryption.md [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+##### Infisical
+[Infisical](https://infisical.com) is an open source alternative to Vault with a unique feature that allows the user to have different values for secrets in different environments (development and production). It can also be used to inject secrets during deployment.
+
+##### No third party
+For this option, it would have to be decided that the satellite config is secure enough to keep the master key in.
+The only difference from the Key Management System section above is that we don't use Google Secrets Manager or any other third party for storing the master key.
+
+## Open Questions
+* How to securely configure the satellite with the encryption master key or credentials to access Secrets Manager?
```
Comment (mobyvb) [inline 20240129-satellite-managed-encryption.md:160]: I think this one is not an open question anymore

File: 20240129-satellite-managed-encryption.md (hunk 13 above)
Comment (wilfred-asomanii) [reply 20240129-satellite-managed-encryption.md]: the end of the paragraph says _These changes are technically unnecessary if the UI is updated to not provide raw API keys for satellite-managed-encryption projects_. Should more emphasis be made?

Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: I guess that's fine - as long as it's mentioned. I think it would help to at least emphasize it (e.g. bold). I guess I just want it to be very obvious to anyone reading this document that we may not make _any_ changes to Uplink and bindings as part of this design.

File: 20240129-satellite-managed-encryption.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+#### Satellite UI Updates
+
+When creating a project, the user should be provided the option to create different types of projects:
+
+* satellite-managed-encryption project
+* user-managed-encryption project
+
+The user-managed-encryption project will be the same as before, and user will always be prompted to enter encryption passphrase when necessary, e.g. when using the file browser or generating an access grant.
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:92]: Yeah, technically they could be separate. The main reason they are related is because we decided that they were - they both are usability improvements, and they both involve updating the access encryption rules.
In the end, we will still want to present a binary decision to the user - either you select the user-managed encryption project that behaves like current projects (user provided passphrase + path encryption). Or you select the satellite-managed encryption project (satellite managed passphrase + no path encryption).
Behind the scenes, we can configure these two things separately, allowing us to toggle them independently in the future. But for the purposes of this design, they will be grouped together in the UI regardless of how it's configured behind the scenes.

Does that make sense? After thinking about your question, I do think it makes sense to have two separate DB flags, but I still think within the UI we should not worry about making that distinction for the initial work.

[the remark this answers is not in the document]
Comment (cam-a) [reply 20240129-satellite-managed-encryption.md:92]: If it's a product decision to package them together then that's good enough explanation for me.

**Review by cam-a — APPROVED** [review]
lgtm

File: 20240129-satellite-managed-encryption.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:93]: well I talked with Wilfred about it, and we actually think splitting the flags is a good idea, since it gives us more options in the future. So it was a good suggestion and we've integrated it into the doc - thanks!
We will still be grouping the two features together from a product perspective, but we do have the option now to, for example, have user-managed encryption but without path encryption.

**Review by jtolio — APPROVED** [review]
nice!

**Comment by elek** [comment]
Ups, looks like I forgot to submit my questions what I added last week (just clicked to the button). Just let me make it clear here: these are not blocking questions, more like some interest about the details, and some more colors :sun_with_face:

File: 20240129-satellite-managed-encryption.md [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+Alternatively, we can use the [GCP Secret Store CSI driver](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp) to mount the Master Key in the Satellite pod without having to expose any credentials at all.
+However, there are [security considerations](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp?tab=readme-ov-file#security-considerations) we have to make.
+
+#### Satellite Project Updates
+
+A new column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and `null` for others.
+Projects that are created before this feature is deployed and projects created after but opt out should have `null` in this column.
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the KMS service generates a cryptographically random passphrase and encrypts it using the master key
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: We didn't define these details, but it would be easy enough to add. How does AESGCM sound for the algorthim? And how does using the existing project salt in the projects table sound for using a salt?

File: 20240129-satellite-managed-encryption.md [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+The file browser in the UI should generate access credentials using the passphrase from the satellite KMS, and no path encryption.
+
+#### Uplink Updates
+
+New versions of Uplink can be updated to check for the new flag on the project, and automatically create accesses with the correct rules during commands like `uplink setup` and `uplink access create`.
+
+This will allow new Uplink versions to properly handle "raw API keys" (access keys without information like satellite or encryption embedded) for "satellite-managed-encryption" projects.
+
+These changes are technically unnecessary if the UI is updated to not provide raw API keys for satellite-managed-encryption projects. However, it may be useful to make these changes to Uplink so that future versions of Uplink can use raw API keys even for satellite-managed-encryption projects.
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: It is not really useful. We moved this to "out of scope" and most likely will not spend effort on this for the time being, since it is unnecessary with the other changes.

File: 20240129-satellite-managed-encryption.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+### Observability
+
+We should ensure that we update analytics events during project creation to identify whether a project was created with satellite-managed or user-managed encryption.
+
+### Test plan
+
+The most important things to test:
+* if a project is created with satellite-managed encryption, it should be very difficult or impossible for a user to unintentionally upload files with path encryption enabled, or with an encryption passphrase other than the project's satellite-managed passphrase.
+* Satellite KMS for encryption passphrases is secure - should be encrypted, and access should be highly protected. Secure transfer of credentials to the client should only happen for project members.
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:121]: I am not sure I understand what you're saying. Any credentials that will be used to access a user's files will ultimately originate from the Satellite UI, even if it is on Edge.
And also, tools outside of the Satellite UI will be able to take advantage of the same features (Edge will be able to do lexicographical ordering, and Uplink will not require passphrase on setup)

File: 20240129-satellite-managed-encryption.md [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+* satellite-managed-encryption project
+* user-managed-encryption project
+
+The user-managed-encryption project will be the same as before, and user will always be prompted to enter encryption passphrase when necessary, e.g. when using the file browser or generating an access grant.
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
+
+The user should not be able to directly retrieve a "raw API key" from the Satellite UI. This prevents the user from unintentionally creating "invalid" credentials using a raw API key with Uplink. The only credentials that can be generated from the access page for a satellite-managed-encryption project should be credentials that already have embedded encryption rules.
+
+The file browser in the UI should generate access credentials using the passphrase from the satellite KMS, and no path encryption.
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:97]: We could, but is it necessary? If the UI is designed to always add the necessary restrictions (encryption rules) to any raw API key before providing it to the user.
Maybe a user could still get the raw key and modify the encryption info if they knew what they were doing, but it would not be possible to do accidentally, I think.
What is the advantage of adding a new DRPC service in your opinion?

File: 20240129-satellite-managed-encryption.md [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+This secrets manager was chosen because it is just as secure as the alternatives, if not more and yet cheaper and easier to manage. It also is more convenient to use as we already use GCP.
+* The KMS service - This is part of the Satellite that will be responsible for fetching the master key at startup, encrypting and decrypting passphrases, and storing and retrieving same from the passphrase store. The master key will be kept in memory throughout the lifetime of the Satellite.
+The satellite will be configured with credentials needed to access the Secret Manager (the master key store).
+* The Passphrase store - This where the encrypted passphrases will be stored. It will be a new column, `passphrase_enc` on the `projects` table. As added benefits to storing this on `projects`, we will be able to back up encrypted passphrases and also rotate master keys.
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md]: Yeah, the "added benefits" is about adding it as an encrypted column on the Satellite DB as opposed to deterministically deriving the project secret from the master key + some unique project data.

**Option 1 (hard to rotate master key)**

* `projectPassphrase = deriveProjectPassphrase(masterKey, projectID)`

If master key needs to change for any reason, it is impossible to rotate it without completely unencrypting/reencrypting each file.

**Option 2 (easy to rotate master key)**

* `projectPassphrase = decrypt(projectPassphraseEnc, masterKey)`

If master key needs to change for any reason, it is easy to rotate it by simply unencrypting/reencrypting the `projects.passphrase_enc` column without needing to change the underlying encryption passphrase.

File: 20240129-satellite-managed-encryption.md [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+
+#### Key Management Store
+
+We need a secure way to store and look up project encryption passphrases on the Satellite. Someone with normal Satellite DB access should not be able to view this data, and ideally it is encrypted. At the same time, it must be very reliable, as losing data about project encryption passphrases means that data may have been lost. The primary features and concerns we should keep in mind:
+* as simple as possible from an implementation perspective, e.g. key-value lookup for project ID -> encryption passphrase
+* secure - should be encrypted, and highly restricted who/what has access to it
+* reliable - should have the same standards for uptime/data loss as metainfo DB
+
+The KMS will be made up of 3 components:
+* The master key and store - The master key will be used to encrypt and decrypt passphrases. The store will be used to securely store the master key. The store will be [Google's Secret Manager](https://cloud.google.com/security/products/secret-manager), which is a secure store for secrets such as API keys and other sensitive data.
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply 20240129-satellite-managed-encryption.md:53]: I agree that it should be easy to have this feature enabled even without an integration with a third party like Google Secret Manager. We can update the document to be more explicit about this, but I think it will naturally happen anyway because we will want to easily test this feature in storj-up without a lot of additional configuration

**Comment by mobyvb** [comment]
@elek I am going to go ahead and merge due to the two approvals and not wanting this to become stale.
I have responded to all your comments, and left our remaining conversations on this PR unresolved, in case you want to respond.
If you suggest changes, we can follow up with a smaller PR. If you do not respond within a few days, I will likely resolve these remaining conversations.
